### PR TITLE
Minipool Rewards Calculator

### PR DESCRIPTION
**Pull request by emersoncloud**
First iteration on a page to estimate rewards for running a minipool with GoGoPool. 

This first pass has a table that shows GGP rewards based on different collateralization percentages. And a table with existing node operators and their stats. 

Thing I still want to add
- [x] Input variable amount of GGP collateral
- [x] Input more than one minipool, and see how max GGP collateral is affected
- [x] Update Node Op table to reflect newly calculated minipool values
- [x] Allow users to see and adjust protocol parameters
- [x] Differentiate investors visually in the table

### CODE REVIEW

File: components/Calculator.tsx [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+  const [avaxAmount, setAvaxAmount] = useState<BigNumber>(BigNumber.from(1000));
+
+  const nodeOpPot = BigNumber.from("45749504487698707175523");
+  const investorPot = BigNumber.from("5083278276410967463947");
+
+  const { data: stakers } = useStakers();
+  const { data: minSeconds } = useGetRewardsEligibilityMinSeconds();
+  const { data: ggpPriceInAVAX } = useGetGGPPriceInAVAX();
+
+  if (!stakers || !minSeconds || !ggpPriceInAVAX) return <Space></Space>;
```
Comment (Namaskar-1F64F) [inline components/Calculator.tsx:26]:
```suggestion
  if (!stakers || !minSeconds || !ggpPriceInAVAX) return null;
```

Space is for putting spaces between elements

File: components/Calculator.tsx [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+  const d = new Date();
+  const epoch = BigNumber.from(Math.floor(d.getTime() / 1000));
```
Comment (Namaskar-1F64F) [inline components/Calculator.tsx:29]:
```suggestion
  const epoch = BigNumber.from(Math.floor(new Date().getTime() / 1000));
```

File: components/Calculator.tsx [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+  let stakeArray: BigNumber[] = [];
+  let tegs = BigNumber.from(0);
```
Comment (Namaskar-1F64F) [inline components/Calculator.tsx:32]: what the hell is a `teg`

File: components/Calculator.tsx [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+      return (
+        toWei(staker.avaxValidatingHighWater) && staker.rewardsStartTime.gt(0)
+        // elapsedSeconds.gte(minSeconds)
+      );
+    })
+    .sort((a, b) => toWei(b.ggpStaked) - toWei(a.ggpStaked))
+    .map((staker) => {
+      // calculate effective ggp stake and total eligible ggp staked
+      const max = staker.avaxValidatingHighWater
+        .mul(BigNumber.from("1500000000000000000"))
```
Comment (Namaskar-1F64F) [inline components/Calculator.tsx:48]: use ether's `parseUnits`

File: components/Calculator.tsx [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+      const ggpAsAVAX = staker.ggpStaked
+        .mul(ggpPriceInAVAX.price)
+        .div(weiValue);
+
+      const effectiveGGPStaked = ggpAsAVAX.gt(max)
+        ? max.mul(weiValue).div(ggpPriceInAVAX.price)
+        : staker.ggpStaked;
+
+      if (staker.stakerAddr != "0xFE5200De605AdCB6306F4CDed77f9A8D9FD47127") {
```
Comment (Namaskar-1F64F) [inline components/Calculator.tsx:59]: this should be an array of investors and do something like

`INVESTOR_WALLETS.includes(staker.stakerAddr)`

which would be in a function called

`isInvestorWallet(addr)`

File: components/Calculator.tsx [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    };
+  });
+
+  const baseColumns = [
+    {
+      title: "GGP Stake",
+      dataIndex: "ggpStake",
+      key: "ggpStake",
+      render: (ggpStake: string) => (
+        <>{`${commify((+formatEther(ggpStake)).toFixed(2))}`}</>
```
Comment (Namaskar-1F64F) [inline components/Calculator.tsx:104]: oh my god i've never seen so many symbols this close to each other

File: components/Calculator.tsx [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+  ];
+
+  const stakerColumns = [
+    { title: "Staker Addr", dataIndex: "stakerAddr", key: "stakerAddr" },
+    ...baseColumns,
+  ];
+
+  let rewardAmounts = [
+    {
+      collat: "10%",
```
Comment (Namaskar-1F64F) [inline components/Calculator.tsx:157]: nit: prefer full names of variables

**Review by Namaskar-1F64F — APPROVED** [review]
Thanks for the PR. All looks good to me.

style wise and functionality I like it. excited to see the new features as well.

thanks for fixing the types 

1. rename percent of all stake (i dunno what it means)
2. collateral perc -> ggp collateralization
3. change addresses to the copyable/clickable emoji versions to keep consistent with the app styling
4. add some description to explain what this does
5. make the investor stand out and call out they are nerfed (make their address a different color and have an antd tooltip that says, investor with reduced rewards or something)

would be happy to get this out and make a follow up as well if it makes sense.

**Review by RusseII — CHANGES_REQUESTED** [review]
Could this be combined with the node operator table? 
CleanShot 2023-06-11 at 11 24 08@2x


Also, the GGP to AVAX price seems wrong 
CleanShot 2023-06-11 at 11 27 44@2x

Percent of all stake for the investors is wrong
CleanShot 2023-06-11 at 11 28 13@2x

Is the plan to create an ROI calculator with inputs? Similar to https://www.rocketpooltool.com/ ? The current rewards box seems kinda confusing compared to inputting my own values

File: components/Calculator.tsx (hunk 3 above)
Comment (emersoncloud) [reply components/Calculator.tsx]: Total Eligible GGP Staked. I'll work on a better name

**Comment by jacktrusler** [comment]
Big mongo PR, all of the changes I worked on had to do with the calculator, it's split into 6 files all located in the `src/components/calculator/` folder, and the page is in `src/pages/calculator.tsx`. Most of the calculations and values are determined in `Calculator.tsx`

 
Screenshot 2023-07-13 at 5 24 44 PM
The main component is the calculator, and it effects the rest of the page with the variables you set. Thats in `YourMinipool.tsx` and `ProtocolSettings.tsx` Protocol settings is that little box at the top that allows you to manually change the GGP price. `YourMinipoolResults.tsx` is self explanatory. 

Screenshot 2023-07-13 at 5 24 44 PM
The first table is the how the calculator value affects your rewards `RatioRewardsTable.tsx`, the second is retail node operators.

Screenshot 2023-07-13 at 5 25 05 PM
And the third is investor node operators, both retail and investors are in `NodeOpRewardTable.tsx`

**Review by Namaskar-1F64F — APPROVED** [review]
https://share.cleanshot.com/jhQw2cVb

i think this is good to merge though. looks good and seems pretty good at all sizes

**Comment by gatesyp** [comment]
> what does this mean exactly? “In addition you’ll earn $1,145.47 from the current* GGP base APY of 7.50%.
what does ggp base apy mean? the current wording seems v confusing

how about something like
“By running a minipool, you will also earn AVAX staking rewards at the current AVAX network APY of 7.50%, valued at $1,145.47” 





--- 

change to "Effective GGP Stake" (possibly add a tooltip to it, to explain what effective stake means? ) 
CleanShot 2023-07-14 at 12 56 24@2x

---- 


a user was confused what the difference between panopticon and this calculator was. not sure if there is actionable feedback here. 

he also didnt notice the "add your mninipool" checkmark, and didnt know that the "Retail Node Ops" table was showing the FUTURE state of protocol after you sign up. 

i recommend changing the header "Retail Node Ops" to something more descriptive , or maybe adding some flavor text under the header to describe that the below tables show the FUTURE state of protocol

**Comment by gatesyp** [comment]
also under Investor Node Ops, changethe "GGP Staked" to "Effective GGP Staked"

**Comment by jacktrusler** [comment]
> “By running a minipool, you will also earn AVAX staking rewards at the current AVAX network APY of 7.50%, valued at $1,145.47”

Not exactly true is the problem, the GGP APY is different than the AVAX network APY for the base rate

**Comment by gatesyp** [comment]
the main thing is you can see:
1. how much GGP rewards you will get next month
2. estimate of annual earnings
3. compared to not using GGP

mission accomplished


^^ from a user

**Comment by jacktrusler** [comment]
I pushed some changes: 

- Changed copy for GGP Minipool to say "In addition you'll earn $1,064.94 from the Avalanche Network at an APY of 7.50%".
- General styling cleanup for different screen sizes.
- Headers of table say "monthly" for rewards.
- Fixed most of the feedback from @Namaskar-1F64F

**Comment by Namaskar-1F64F** [comment]
@jacktrusler  is it supposed to be so

**Comment by jacktrusler** [comment]
Yea I thought it was too stretched out across the page

**Comment by Namaskar-1F64F** [comment]
Thanks for the contribution! @emersoncloud @jacktrusler

As first-time contributors, you will receive 3 months of dappling pro, on us. Let's keep growing 🌱🙏

**Pull request merged**